internal/pubsub/awspubsub/awspubsub.go

```diff
@@ -0,0 +1,202 @@
+func OpenTopic(ctx context.Context, client *sns.SNS, topicARN string, opts *TopicOptions) *pubsub.Topic {
+	dt := openTopic(ctx, client, topicARN)
+	return pubsub.NewTopic(dt)
+}
+
+// openTopic returns the driver for OpenTopic. This function exists so the test
+// harness can get the driver interface implementation if it needs to.
+func openTopic(ctx context.Context, client *sns.SNS, topicARN string) driver.Topic {
+	return &topic{client: client, arn: topicARN}
+}
+
+// SendBatch implements driver.Topic.SendBatch.
+func (t *topic) SendBatch(ctx context.Context, dms []*driver.Message) error {
+	for _, dm := range dms {
+		attrs := map[string]*sns.MessageAttributeValue{}
+		for k, v := range dm.Metadata {
+			attrs[k] = &sns.MessageAttributeValue{
+				DataType:    aws.String("String"),
+				StringValue: aws.String(v),
+			}
+		}
+		params := sns.PublishInput{
+			Message:           aws.String(string(dm.Body)),
+			MessageAttributes: attrs,
+			TopicArn:          &t.arn,
+		}
+		if err := t.send(&params); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (t *topic) send(params *sns.PublishInput) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	req, _ := t.client.PublishRequest(params)
+	return req.Send()
+}
+
+// IsRetryable implements driver.Topic.IsRetryable.
+func (t *topic) IsRetryable(error) bool {
+	// The client handles retries.
+	return false
+}
+
+// As implements driver.Topic.As.
+func (t *topic) As(i interface{}) bool {
+	c, ok := i.(**sns.SNS)
+	if !ok {
+		return false
+	}
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	*c = t.client
+	return true
+}
+
+type subscription struct {
+	mu     sync.Mutex
+	client *sqs.SQS
+	qURL   string
+}
+
+// SubscriptionOptions will contain configuration for subscriptions.
+type SubscriptionOptions struct{}
+
+func OpenSubscription(ctx context.Context, client *sqs.SQS, qURL string, opts *SubscriptionOptions) *pubsub.Subscription {
+	ds := openSubscription(ctx, client, qURL)
+	return pubsub.NewSubscription(ds)
+}
+
+// openSubscription returns a driver.Subscription.
+func openSubscription(ctx context.Context, client *sqs.SQS, qURL string) driver.Subscription {
+	return &subscription{client: client, qURL: qURL}
+}
+
+// ReceiveBatch implements driver.Subscription.ReceiveBatch.
+func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
+	output, err := s.receiveMessages()
+	if err != nil {
+		return nil, err
+	}
+	var ms []*driver.Message
+	for _, m := range output.Messages {
+		type Attribute struct {
+			Value string
+		}
+		type MsgBody struct {
+			Message           string
+			MessageAttributes map[string]Attribute
+		}
+		var body MsgBody
+		if err := json.Unmarshal([]byte(*m.Body), &body); err != nil {
+			return nil, err
+		}
+		attrs := map[string]string{}
+		for k, v := range body.MessageAttributes {
+			attrs[k] = v.Value
+		}
+		m := &driver.Message{
+			Body:     []byte(body.Message),
+			Metadata: attrs,
+			AckID:    *m.ReceiptHandle,
```

Comment:
nit: there's no need to dereference it, is there?

Reply:
Apparently there's not. Done.